src/include/mpir_gpu_util.h

```diff
@@ -0,0 +1,82 @@
+/*
+ * Copyright (C) by Argonne National Laboratory
+ *     See COPYRIGHT in top-level directory
+ */
+
+#ifndef MPIR_GPU_UTIL_H_INCLUDED
+#define MPIR_GPU_UTIL_H_INCLUDED
+
+/* If buf is device memory, allocate a host buffer that can hold the data,
+ * return the host buffer, or NULL if buf is not device memory */
+MPL_STATIC_INLINE_PREFIX void *MPIR_gpu_host_alloc(const void *buf,
+                                                   MPI_Aint count, MPI_Datatype datatype)
+{
+    MPL_pointer_attr_t attr;
+    MPIR_GPU_query_pointer_attr(buf, &attr);
+
+    if (attr.type != MPL_GPU_POINTER_DEV) {
+        return NULL;
+    } else {
+        MPI_Aint extent, true_lb, true_extent;
+        MPIR_Datatype_get_extent_macro(datatype, extent);
+        MPIR_Type_get_true_extent_impl(datatype, &true_lb, &true_extent);
+
+        MPI_Aint shift = true_lb;
+        MPI_Aint size = true_extent;
+        if (count > 1) {
+            if (extent >= 0) {
+                size += extent * (count - 1);
+            } else {
+                MPI_Aint size_before = (-extent) * (count - 1);
+                /* extra counts are being packed *before*, so need *increase* the size */
+                size += size_before;
+                /* the allocated pointer will be *already* shifted by this amount */
+                shift -= size_before;
+            }
+        }
+        void *host_buf = MPL_malloc(size, MPL_MEM_OTHER);
+        MPIR_Assert(host_buf);
+
+        host_buf = (char *) host_buf - shift;
+        return host_buf;
```

Comment:
@raffenet Thought about it a bit more. The previous fix didn't account for the case when `extent` is negative, which will put the extra count data *before* the first data, thus we may need to allocate extra size to account for.

Could you review this commit again? Meanwhile, I'll rebase and test it in PR #5000